Sources/SwiftMemcache/MemcachedConnection.swift

```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-memcache-gsoc open source project
+//
+// Copyright (c) 2023 Apple Inc. and the swift-memcache-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-memcache-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+@_spi(AsyncChannel)
+
+import NIOCore
+import NIOPosix
+
+public actor MemcachedConnection {
+    private let channel: NIOAsyncChannel<ByteBuffer, ByteBuffer>
+    private let bufferAllocator: ByteBufferAllocator
+    private let group: MultiThreadedEventLoopGroup
+    
+    public init(host: String, port: Int, group: MultiThreadedEventLoopGroup) async throws {
+            self.group = group
+            let bootstrap = ClientBootstrap(group: self.group)
+                .channelInitializer { channel in
+                    return channel.pipeline.addHandlers([MessageToByteHandler(MemcachedRequestEncoder()), ByteToMessageHandler(MemcachedResponseDecoder())])
+                }
+
+            let rawChannel = try await bootstrap.connect(host: host, port: port).get()
+            self.channel = try NIOAsyncChannel(synchronouslyWrapping: rawChannel, inboundType: ByteBuffer.self, outboundType: ByteBuffer.self)
+        
+        self.bufferAllocator = ByteBufferAllocator()
+    }
+
+
+    public func get(_ key: String) async throws -> ByteBuffer? {
+        var flags = MemcachedFlags()
+        flags.shouldReturnValue = true
+        let command = MemcachedRequest.GetCommand(key: key, flags: flags)
+        let request = MemcachedRequest.get(command)
+
+        var buffer = self.bufferAllocator.buffer(capacity: 0)
+        let encoder = MemcachedRequestEncoder()
+        try encoder.encode(data: request, out: &buffer)
+        try await self.channel.outboundWriter.write(buffer)
+
+        let responseBuffer = try await self.channel.inboundStream.first(where: { _ in true })
```

Comment:
We can't do `.first` here. The problem is that `first()` creates an iterator to the `inboundStream`. The `inboundStream` allows only a single iterator and when we deinit the iterator the `inboundStream` will terminate. This is going to show one of the major architectural changes that we have to make to get this to work. We can only ever create a single iterator but we have multiple methods that want to create requests and then await the response. We cannot store the iterator as a property because we will end up with this error:

```
actor Foo {
    private var iterator: AsyncStream<Void>.Iterator

    init(iterator: AsyncStream<Void>.Iterator) {
        self.iterator = iterator
    }

    func foo() async {
        self.iterator.next() // Error: Cannot call mutating async function 'next()' on actor-isolated property 'iterator'
    }
}
```

The complicated work here would be to dispatch all requests into an `AsyncStream` which we consume and handle in order and create a bunch of continuations to suspend the calling tasks. Additionally, we would need to handle task cancellation. For now I would recommend to ignore task cancellation and just do the async stream work. This means we have to create an `AsyncStream<(MemcachedRequest, CheckedContinuation<MemcachedResponse, Error>)>` in our init. Add a new `run()` method to this actor where we consume the async stream via a `for await (request, continuation) in stream`. Lastly, our `get` and `set` method need to create the request and a continuation via `withCheckedThrowingContinuation` and yield it to the async streams continuation.

You can use this helper code to easily create yourself an async stream:

```
extension AsyncStream {
  fileprivate static func makeStream(
      of elementType: Element.Type = Element.self,
      bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded
  ) -> (stream: AsyncStream<Element>, continuation: AsyncStream<Element>.Continuation) {
    var continuation: AsyncStream<Element>.Continuation!
    let stream = AsyncStream<Element>(bufferingPolicy: limit) { continuation = $0 }
    return (stream: stream, continuation: continuation!)
  }
}

```